calico_upgrade/Makefile

```diff
@@ -0,0 +1,326 @@
+	docker build -t $(UPGRADE_CONTAINER_NAME) -f pkg/Dockerfile.calico_upgrade .
+	touch $@
+
+## Build calico-upgrade
+binary: $(CALICO_UPGRADE_FILES) vendor
+	# Don't try to "install" the intermediate build files (.a .o) when not on linux
+	# since there are no write permissions for them in our linux build container.
+	if [ "$(OS)" == "linux" ]; then \
+		INSTALL_FLAG=" -i "; \
+	fi; \
+	GOOS=$(OS) GOARCH=$(ARCH) CGO_ENABLED=0 go build -v $$INSTALL_FLAG -o dist/calico-upgrade-$(OS)-$(ARCH) $(LDFLAGS) "./pkg/calicoupgrade.go"
+
+dist/calico-upgrade: $(CALICO_UPGRADE_FILES) vendor
+	$(MAKE) dist/calico-upgrade-linux-amd64
+	mv dist/calico-upgrade-linux-amd64 dist/calico-upgrade
+
+dist/calico-upgrade-linux-amd64: $(CALICO_UPGRADE_FILES) vendor
+	$(MAKE) OS=linux ARCH=amd64 binary-containerized
+
+dist/calico-upgrade-darwin-amd64: $(CALICO_UPGRADE_FILES) vendor
+	$(MAKE) OS=darwin ARCH=amd64 binary-containerized
+
+dist/calico-upgrade-windows-amd64.exe: $(CALICO_UPGRADE_FILES) vendor
+	$(MAKE) OS=windows ARCH=amd64 binary-containerized
+	mv dist/calico-upgrade-windows-amd64 dist/calico-upgrade-windows-amd64.exe
+
+## Run the build in a container. Useful for CI
+binary-containerized: $(CALICO_UPGRADE_FILES) vendor
+	mkdir -p dist
+	-mkdir -p .go-pkg-cache
+	docker run --rm \
+	  -e OS=$(OS) -e ARCH=$(ARCH) \
+	  -e CALICO_UPGRADE_VERSION=$(CALICO_UPGRADE_VERSION) \
+	  -e CALICO_UPGRADE_BUILD_DATE=$(CALICO_UPGRADE_BUILD_DATE) -e CALICO_UPGRADE_GIT_REVISION=$(CALICO_UPGRADE_GIT_REVISION) \
+	  -v $(CURDIR):/go/src/$(PACKAGE_NAME):ro \
+	  -v $(CURDIR)/dist:/go/src/$(PACKAGE_NAME)/dist \
+    -e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+    -v $(CURDIR)/.go-pkg-cache:/go/pkg/:rw \
+	  $(CALICO_BUILD) sh -c '\
+	    cd /go/src/$(PACKAGE_NAME) && \
+	    make OS=$(OS) ARCH=$(ARCH) \
+	         CALICO_UPGRADE_VERSION=$(CALICO_UPGRADE_VERSION)  \
+	         CALICO_UPGRADE_BUILD_DATE=$(CALICO_UPGRADE_BUILD_DATE) CALICO_UPGRADE_GIT_REVISION=$(CALICO_UPGRADE_GIT_REVISION) \
+	         binary'
+
+.PHONY: install
+install:
+	CGO_ENABLED=0 go install $(PACKAGE_NAME)/calico_upgrade
+
+###############################################################################
+# calico-upgrade UTs
+###############################################################################
+.PHONY: ut
+## Run the Unit Tests locally
+ut: dist/calico-upgrade
+	# Run tests in random order find tests recursively (-r).
+	ginkgo -cover -r --skipPackage vendor pkg/*
+
+	@echo
+	@echo '+==============+'
+	@echo '| All coverage |'
+	@echo '+==============+'
+	@echo
+	@find ./pkg/ -iname '*.coverprofile' | xargs -I _ go tool cover -func=_
+
+	@echo
+	@echo '+==================+'
+	@echo '| Missing coverage |'
+	@echo '+==================+'
+	@echo
+	@find ./pkg/ -iname '*.coverprofile' | xargs -I _ go tool cover -func=_ | grep -v '100.0%'
+
+PHONY: test-containerized
+## Run the tests in a container. Useful for CI, Mac dev.
+test-containerized: dist/calico-upgrade
```

Comment:
This fails for me because the Makefile tries to read something (versions.yaml) outside the directory that we mount into the containser